File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        og_clone = defaultdict(lambda: None)
        def dfs(node: Optional['Node']):
            if node == None:
                return None
            if og_clone[node] != None:
                return og_clone[node]
            new_node = Node(node.val)
            og_clone[node] = new_node
            children = []
            for child in node.neighbors:
                child_clone = dfs(child)
                children.append(child_clone)
            new_node.neighbors = children
            return new_node
        return dfs(node)

File: 133-clone-graph/clone-graph.cpp
class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<int, Node*> clone_map;
        return dfs(node, clone_map);
    }

    Node* dfs(Node* node, unordered_map<int, Node*>& clone_map) {
        if(node == nullptr) return nullptr;
        Node* node_clone = new Node(node->val);
        clone_map[node->val] = node_clone;
        for(Node* node_child: node->neighbors){
            if(!clone_map.contains(node_child->val)){
                node_clone->neighbors.push_back(dfs(node_child, clone_map));
            } else {
                node_clone->neighbors.push_back(clone_map[node_child->val]);
            }
        }
        return node_clone;
    }
};